app/src/main/java/fi/aalto/legroup/achso/entities/PooledVideo.java

```diff
@@ -0,0 +1,132 @@
+     *                            allocate when inflated.
+     */
+    public PooledVideo(int annotationCountHint, boolean hasLocationHint) {
+        video = new Video();
+        reserveAnnotations(annotationCountHint);
+        if (hasLocationHint) {
+            location = new Location("pooled");
+        }
+    }
+
+    private void reserveAnnotations(int count) {
+
+        if (annotations != null) {
+            annotations.ensureCapacity(count);
+        } else {
+            annotations = new ArrayList<>(count);
+        }
+
+        // Pre-create annotation objects for the
+        while (annotations.size() < count) {
+            Annotation annotation = new Annotation();
+
+            // Annotations are expected to have a position ready so we don't need to create
+            // new position objects all the time
+            annotation.setPosition(new PointF());
+
+            annotations.add(annotation);
+        }
+    }
+
+    /**
+     * How many annotations can be stored without having to allocate.
+     */
+    public int getAnnotationCapacity() {
+        return annotations.size();
+    }
+
+    /**
+     * Is the Video object of this pooled instance still in use.
+     */
+    public boolean isInUse() {
+        return inUse;
+    }
+
+    /**
+     * Create the real Video object with the specified annotation count.
+     * Note: You should call free() after you are done with the video object.
+     *
+     * @param annotationCount How many annotations should the video contains. Does not allocate
+     *                        if `annotationCount` is less or equal than
+     *                        `getAnnotationCapacity()`.
+     *                        Otherwise expands the annotation capacity to fit.
+     * @param hasLocation     Should the returned Video have it's location field initialized, if
+     *                        set to true the location is initialized into the persistent location
+     *                        of this pool.
+     */
+    public Video create(int annotationCount, boolean hasLocation) {
+
+        if (inUse) {
+            // Fallback, this shouldn't happen often
+            Log.w(TAG, "Created a new Video object from pool (old wasn't released)");
+            return new PooledVideo(annotationCount, hasLocation).create(annotationCount,
+                    hasLocation);
+        }
+
+        reserveAnnotations(annotationCount);
+        List<Annotation> subAnn = annotations.subList(0, annotationCount);
+        video.setAnnotations(subAnn);
+        if (hasLocation) {
+            if (location == null) {
+                location = new Location("dynamic pooled");
+            }
+            video.setLocation(location);
+        } else {
+            video.setLocation(null);
+        }
+        return video;
+    }
+
+    /**
+     * Mark that the Video can be re-used from this pool safely again.
+     */
+    public void free() {
```

Comment:
I think `release` might be a more conventional name in Java. It might describe this slightly better than `free`, since this doesn’t actually free any memory.